util/sqlkiller/sqlkiller.go

```diff
@@ -0,0 +1,85 @@
+// Copyright 2023 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package sqlkiller
+
+import (
+	"fmt"
+	"sync/atomic"
+
+	"github.com/pingcap/tidb/util/dbterror/exeerrors"
+	"github.com/pingcap/tidb/util/logutil"
+	"go.uber.org/zap"
+)
+
+// Kill reasons.
+const (
+	QueryInterrupted uint32 = iota + 1
+	MaxExecTimeExceeded
+	QueryMemoryExceeded
+	ServerMemoryExceeded
+)
+
+// SQLKiller is used to kill a query.
+type SQLKiller struct {
+	status *uint32
+	connid uint64
+}
+
+// NewSQLKiller creates a new SqlKiller.
+func NewSQLKiller(connid uint64, status *uint32) *SQLKiller {
+	return &SQLKiller{connid: connid, status: status}
+}
+
+// SendKillSignal sends a kill signal to the query.
+func (killer *SQLKiller) SendKillSignal(reason uint32) {
```

Comment:
might return bool value to indicate if send successful?

Reply:
I don't think it's necessary to return if send was successful. Only when a kill signal succeeds in sending, other kill signals may fail to send. In this case, the SQL is killed regardless of whether subsequent signals are successfully sent.